File: samples/datastruct/java/stack.java
class stack {
	private structure up;
	private int size;

	public stack() {
		init();
	}

	public void init() {
		up = null;
		size = 0;
	}

	public int size() {
		return size;
	}

	public boolean push(int item) {
		structure tempNode = new structure(item, null, null);

		if (tempNode == null)
			return false;

		size++;

		if (up == null)
			up = tempNode;
		else {
			up.setPrev(tempNode);
			tempNode.setNext(up);
			up = tempNode;
		}
			
		return true;
	}	
	
	public boolean pop() {
		size--;
		up = up.getNext();

		return true;
	}	

	public int top() {
		if (up != null)
			return up.getItem();
		else
			return 0;
	}

	public int bottom() {
		if (up != null)
			return up.getItem();
		else
			return 0;
	}
	
	public int length() {
		return size;
	}
}

File: samples/datastruct/c/stack.cpp
/*
 * stack.cpp
 * Vino Crescini
 * 8 December 2000
 */

#include <stdlib.h>
#include "stack.h"

stackType::stackType() 
{
  init();
}

stackType::~stackType()
{
  // cleanup
  int tempInt;
  while (pop(&tempInt)); 
}

void stackType::init()
{
  stack = NULL;
  size  = 0;
}

int stackType::length()
{
  return size;
}

bool stackType::push(int item)
{
  structure *tempNode = new structure();

  if (tempNode) {
    size++;
    tempNode->setStructure(item, stack, NULL);
    stack = tempNode;	

    return true; 
  }
	
  return false;
}

bool stackType::pop(int *item)
{
  structure *tempNode;

  if (!item || size <= 0) 
    return false;

  size--;
  tempNode = stack;
  *item    = stack->getItem();
  stack	   = stack->getNext();
	
  delete tempNode;

  return true;
}

bool stackType::top(int *item)
{
  if (!item || size <= 0)
    return false;

  *item = stack->getItem();

  return true;
}
